server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/WeightFunction.java

```diff
@@ -13,97 +13,40 @@
 import org.elasticsearch.cluster.routing.RoutingNodes;
 import org.elasticsearch.cluster.routing.allocation.WriteLoadForecaster;
-import org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator.ProjectIndex;
 import org.elasticsearch.index.shard.ShardId;
 
-/**
- * This class is the primary weight function used to create balanced over nodes and shards in the cluster.
- * Currently this function has 3 properties:
- * <ul>
- * <li><code>index balance</code> - balance property over shards per index</li>
- * <li><code>shard balance</code> - balance property over shards per cluster</li>
- * </ul>
- * <p>
- * Each of these properties are expressed as factor such that the properties factor defines the relative
- * importance of the property for the weight function. For example if the weight function should calculate
- * the weights only based on a global (shard) balance the index balance can be set to {@code 0.0} and will
- * in turn have no effect on the distribution.
- * </p>
- * The weight per index is calculated based on the following formula:
- * <ul>
- * <li>
- * <code>weight<sub>index</sub>(node, index) = indexBalance * (node.numShards(index) - avgShardsPerNode(index))</code>
- * </li>
- * <li>
- * <code>weight<sub>node</sub>(node, index) = shardBalance * (node.numShards() - avgShardsPerNode)</code>
- * </li>
- * </ul>
- * <code>weight(node, index) = weight<sub>index</sub>(node, index) + weight<sub>node</sub>(node, index)</code>
- */
-public class WeightFunction {
-
-    private final float theta0;
-    private final float theta1;
-    private final float theta2;
-    private final float theta3;
-
-    public WeightFunction(float shardBalance, float indexBalance, float writeLoadBalance, float diskUsageBalance) {
-        float sum = shardBalance + indexBalance + writeLoadBalance + diskUsageBalance;
-        if (sum <= 0.0f) {
-            throw new IllegalArgumentException("Balance factors must sum to a value > 0 but was: " + sum);
-        }
-        theta0 = shardBalance / sum;
-        theta1 = indexBalance / sum;
-        theta2 = writeLoadBalance / sum;
-        theta3 = diskUsageBalance / sum;
-    }
+public interface WeightFunction {
 
     float calculateNodeWeightWithIndex(
         BalancedShardsAllocator.Balancer balancer,
         BalancedShardsAllocator.ModelNode node,
-        ProjectIndex index
-    ) {
-        final float weightIndex = node.numShards(index) - balancer.avgShardsPerNode(index);
-        final float nodeWeight = calculateNodeWeight(
-            node.numShards(),
-            balancer.avgShardsPerNode(),
-            node.writeLoad(),
-            balancer.avgWriteLoadPerNode(),
-            node.diskUsageInBytes(),
-            balancer.avgDiskUsageInBytesPerNode()
-        );
-        return nodeWeight + theta1 * weightIndex;
-    }
+        BalancedShardsAllocator.ProjectIndex index
+    );
 
-    public float calculateNodeWeight(
+    float calculateNodeWeight(
```

Comment:
It is only this function that would be different depending on if we want per tier function or one for both tiers, right? I wonder if initially we could limit the change to this? Or is that too hacky? That reduces the size and we could potentially follow up with a refactoring.

Reply:
I think `minWeightDelta` _should_ also be specific. That calculation includes `shardWriteLoad`, so the deltas would be different in each tier.

In saying that, the POC only set the forecasted write load on search nodes to zero and left the rest untouched.